### PR TITLE
Allow frontend use of the backend's core hashing functions

### PR DESCRIPTION
**Pull request by bmdoil**
These are a series of backports from PG13 that enable frontend code to utilize the backend's hashing functions.

Recent upstream improvements to pg_dump depend on these hash functions. Also, the backports enable further refactor opportunities for GPDB specific aspects of pg_dump and pg_upgrade. 

48995040d5e7b1e9bac35d72aff326cae002219d simplehash: Allow use of simplehash without MemoryContext.
da41d71070d14ecd9e2f4bbe275c98a136826d4b simplehash: Allow for use in frontend code.
07b95c3d8334f737d4717c91967729f7721e785c Move bitmap_hash and bitmap_match to bitmapset.c.
9341c783cc42ffae5860c86bdc713bd47d734ffd Put all the prototypes for hashfn.c into the same header file.
a91e2fa94180f24dd68fb6c99136cda820e02089 Adapt hashfn.c and hashutils.h for frontend use.
05d8449e73694585b59f8b03aaa087f04cc4679a Move src/backend/utils/hash/hashfn.c to src/common

Discussion: 
http://postgr.es/m/CA+Tgmob8oyh02NrZW=xCScB+5GyJ-jVowE3+TWTUmPF=FsGWTA@mail.gmail.com
http://postgr.es/m/CA+TgmoaRiG4TXND8QuM6JXFRkM_1wL2ZNhzaUKsuec9-4yrkgw@mail.gmail.com

### CODE REVIEW

**Comment by Aegeaner** [comment]
_**Experimenting with hash tables inside pg_dump**_ : https://www.postgresql.org/message-id/flat/2595220.1634855245%40sss.pgh.pa.us 
There is a discussion thread in community about applying simplehash in pg_dump, but it is still not merged into upstream yet?

**Comment by bmdoil** [comment]
> _**Experimenting with hash tables inside pg_dump**_ : https://www.postgresql.org/message-id/flat/2595220.1634855245%40sss.pgh.pa.us There is a discussion thread in community about applying simplehash in pg_dump, but it is still not merged into upstream yet?

The relevant upstream commit is https://github.com/postgres/postgres/commit/92316a4582a5714d4e494aaf90360860e7fec37a

The plan is to backport the above commit along with several others related to pg_dump performance that have been merged upstream from PG13 onward. An upcoming PR will open the discussion on that.

File: src/backend/utils/hash/hashfn.c [hunk 1]
```diff
@@ -701,7 +700,6 @@ uint32_hash(const void *key, Size keysize)
 	Assert(keysize == sizeof(uint32));
 	return DatumGetUInt32(hash_uint32(*((const uint32 *) key)));
 }
-
 /*
```
Comment (bmdoil) [inline src/backend/utils/hash/hashfn.c:699]: This is a GPDB specific no-op with 1 caller, segadmin.c:get_availableDbId()

Can we replace it with uint32_hash?

File: src/common/hashfn.c [hunk 2]
```diff
@@ -699,9 +694,8 @@ uint32
 uint32_hash(const void *key, Size keysize)
 {
 	Assert(keysize == sizeof(uint32));
-	return DatumGetUInt32(hash_uint32(*((const uint32 *) key)));
+	return hash_bytes_uint32(*((const uint32 *) key));
 }
-
 /*
  * int32_hash: hash function for int32: no-op
  */
```
Comment (bmdoil) [inline src/common/hashfn.c:701]: This is a GPDB specific no-op with 1 caller, segadmin.c:get_availableDbId()

Can we replace it with uint32_hash?

**Review by soumyadeep2007 — APPROVED** [review]
Looks good for 7X.

However for the upcoming 6X backport, there are certain changes that make me uncomfortable. For instance, lets say an extension built against 6X imports `string_hash` from `"utils/hsearch.h"`. But now, `Put all the prototypes for hashfn.c into the same header file` has changed it so that `string_hash` is to be imported from `utils/hashutils.h"`. This will break those extensions, if they didn't also include `utils/hashutils.h` incidentally. We might have to reject certain changes for 6X or maybe introduce guarded externs w. `#ifndef` in `utils/hsearch.h` for instance.

**Review by Aegeaner — APPROVED** [review]
LGTM

**Pull request merged**